collector/mdadm_linux.go

```diff
@@ -18,199 +18,141 @@ package collector
 import (
 	"fmt"
 	"io/ioutil"
+	"math"
 	"os"
-	"regexp"
 	"strconv"
 	"strings"
 
+	"syscall"
+	"unsafe"
+
 	"github.com/prometheus/client_golang/prometheus"
 	"github.com/prometheus/common/log"
 )
 
-var (
-	statuslineRE             = regexp.MustCompile(`(\d+) blocks .*\[(\d+)/(\d+)\] \[[U_]+\]`)
-	raid0lineRE              = regexp.MustCompile(`(\d+) blocks .*\d+k chunks`)
-	buildlineRE              = regexp.MustCompile(`\((\d+)/\d+\)`)
-	unknownPersonalityLineRE = regexp.MustCompile(`(\d+) blocks (.*)`)
-	raidPersonalityRE        = regexp.MustCompile(`raid[0-9]+`)
+type (
+	mdStatus struct {
+		name string
+		//active       bool
+		active       float64
+		disksActive  float64
+		disksFailed  float64
+		disksMissing float64
+		disksSpare   float64
+		bytesTotal   float64
+		bytesSynced  float64
+	}
+
+	mdu_array_info struct {
+		// Generic constant information
+		major_version  int32
+		minor_version  int32
+		patch_version  int32
+		ctime          uint32
+		level          int32 //  RAID level
+		size           int32 //  useless - only works for disk sizes that fit in an int32
+		nr_disks       int32 //  Number of physically present disks. (working + failed)
+		raid_disks     int32 //  Number of disks in the RAID.
+		md_minor       int32 //     This may be larger than "nr" if disks are missing
+		not_persistent int32 //     and smaller than "nr" when spare disks are around.
+		// Generic state information
+		utime         uint32 //  Superblock update time
+		state         int32  //  State bits (clean, ...)
+		active_disks  int32  //  Number of active (in sync) disks
+		working_disks int32  //  Number of working disks (active + sync)
+		failed_disks  int32  //  Number of failed disks
+		spare_disks   int32  //  Number of spare (stand-by) disks
+		// Personality information
+		layout     int32 //  the array's physical layout
+		chunk_size int32 //  chunk size in bytes
+	}
 )
 
-type mdStatus struct {
-	name         string
-	active       bool
-	disksActive  int64
-	disksTotal   int64
-	blocksTotal  int64
-	blocksSynced int64
-}
+const (
+	// _IOR definition:
+	// https://github.com/torvalds/linux/blob/e067eba5871c6922539dc1728699c14e6b22590f/include/uapi/asm-generic/ioctl.h#L85
+	// from major.h
+	// #define MD_MAJOR                9
+	// from md_u.h
+	// #define GET_ARRAY_INFO          _IOR (MD_MAJOR, 0x11, mdu_array_info_t)
+
+	GET_ARRAY_INFO = (0x8048 << 16) | 0x0911
```

Comment:
Best bet is probably going to be to expose this constant via `x/sys/unix` and their code generator. This will probably only work on amd64 when hard-coded.